hw/ip/otbn/rtl/otbn_model.cc

```diff
@@ -0,0 +1,141 @@
+#include "svdpi.h"
+
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <fcntl.h>
+#include <sstream>
+#include <iostream>
+#include <string>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+// Candidate for hw/dv/verilator/cpp
+/**
+ * Guard class for SV Scope
+ *
+ * This guard ensures that all functions in the context where it is instantiated
+ * are executed in an SVScope. It will restore the previous scope at destruction
+ * and thereby make sure that the previous scope will be restored for all paths
+ * that exit the scope.
+ */
+class SVScoped {
+  svScope m_prevScope = 0;
+  SVScoped() {}
+
+ public:
+  SVScoped(std::string scopeName) : SVScoped(scopeName.c_str()) {}
+  SVScoped(const char *scopeName) : SVScoped(svGetScopeFromName(scopeName)) {}
+
+  SVScoped(svScope scope) { m_prevScope = svSetScope(scope); }
+  ~SVScoped() { svSetScope(m_prevScope); }
+};
+
+extern "C" {
+
+extern int simutil_verilator_get_mem(int index, const svBitVecVal *val);
+extern int simutil_verilator_set_mem(int index, const svBitVecVal *val);
+
+int run_model(const char *imem_scope, int imem_words, const char *dmem_scope,
+              int dmem_words) {
+  FILE *fp = 0;
```

Comment:
Do not use implicit conversion from `0` or `NULL` into pointers in C++. Always use `nullptr`.

That said, rather than have a single `fp` variable you use everywhere, consider creating a new one with `auto *my_file = fopen();` any time you need to open a file.

Reply:
Sorry, I have a couple of "bad habits" from Verilator (mainly as we support pre-c++11), will get this and other fixed